### ISSUE TITLE
GP Regression with Uncertain Inputs

### ISSUE THREAD

**Issue by sahandrez** [cornellius-gp/gpytorch#659]
Hello, 

I would like to know if GPyTorch currently supports a way to propagate uncertainty during the prediction step (e.g. one possible approach is [this](http://papers.nips.cc/paper/2313-gaussian-process-priors-with-uncertain-inputs-application-to-multiple-step-ahead-time-series-forecasting.pdf)). 

I'm interested in using the GP as a multiple-step predictor, so its output and the predicted uncertainty needs to be used again as inputs. 

Thanks,
Sahand

**Comment by jacobrgardner** [comment]
Sorry for the delay responding!

The answer is "sort of" / "not as a built in first class feature, but something you might be able to accomplish" off the shelf. 

If you sample `s` x values from p(x) and then stick them in a `s x n x d`, (and for training data repeat your y values to be `s x n`), then when you call your model you'll get the loss (in training mode) for each sample that you can mean over, and in eval mode you'll get predictions over each sample s.

The sampling approach will honestly be faster than doing EP or VI with access to a GPU for small to moderate (maybe 5k) data points. Beyond that, you'll probably want to just write a Pyro + GPyTorch model that explicitly places a distribution over x -- we don't have a direct example for this, but I'm thinking it ought to work out of the box.

**Comment by sahandrez** [comment]
Thank you very much for your reply. The sampling method would definitely be the way to implement this then.

**Comment by jejjohnson** [comment]
Hello,

I have the exact problem of trying to put a prior distribution on the inputs but I haven't had any luck with setting the prior to the training data. I was wondering if you could give a bit more intuition where I could find the attribute? 

I was looking through your tutorial [here](https://gpytorch.readthedocs.io/en/latest/examples/09_Pyro_Integration/Pyro_Variational_GP_Regression.html) where you use the Pyro library for the variational distribution. Below is the snippet where I try to set the prior in places that I intuitively thing they should be but I've had no luck.

```python
class PVGPRegressionModel(gpytorch.models.PyroVariationalGP):
    def __init__(self, train_x, train_y, likelihood, name_prefix="mixture_gp"):
        # Define all the variational stuff
        variational_distribution = gpytorch.variational.CholeskyVariationalDistribution(
            num_inducing_points=train_y.numel(),
        )
        variational_strategy = gpytorch.variational.VariationalStrategy(
            self, train_x, variational_distribution
        )
        
        # =============================
        # Set Prior Distribution to X
        # =============================
        
        # Option I - variation distribution
        try:
            variational_distribution.set_prior("train_x", dist.Normal(train_x, 0.1).to_event())
        except AttributeError:
            print('No attribute found in var dist. ...')
            pass
        
        # Option II - variation strategy
        try:
            variational_strategy.set_prior("train_x", dist.Normal(train_x, 0.1).to_event())
        except AttributeError:
            print('No attribute found in var. strat. ...')
            pass

        # Standard initializtation
        super(PVGPRegressionModel, self).__init__(variational_strategy, likelihood, num_data=train_y.numel())
        self.likelihood = likelihood

        # Mean, covar
        self.mean_module = gpytorch.means.ConstantMean()
        self.covar_module = gpytorch.kernels.ScaleKernel(
            gpytorch.kernels.MaternKernel(nu=1.5)
        )

    def forward(self, x):
        mean = self.mean_module(x)  # Returns an n_data vec
        covar = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean, covar)

model = PVGPRegressionModel(train_x, train_y, gpytorch.likelihoods.GaussianLikelihood())

try:
    model.set_prior("X", dist.Normal(X_prior_mean, 0.1).to_event())
except AttributeError:
    print('No prior found in model. ...')
    pass
```

Any insights would be most helpful! Thanks in advanced!
Emmanuel
